### PR TITLE
Add requires_grad argument to `to`.

### PR DESCRIPTION
**Pull request by gchanan**

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -1645,6 +1645,39 @@ def test_to_with_tensor(self):
                 self.assertEqual(a.device, b.to(a).device)
                 self.assertEqual(b.device, a.to(b).device)
 
+    def test_to_with_requires_grad(self):
+        # only requires_grad
+        a = torch.tensor(5.)
+        self.assertEqual(a, a.to(requires_grad=True))
```
Comment (SsnL) [inline test/test_torch.py]: Is this correct? I thought that the returned tensor should be a different object (out-of-place).

Comment (apaszke) [reply test/test_torch.py]: It is a different object, but it is equal (we don't check autograd flags when testing equality)

File: test/test_torch.py [hunk 2]
```diff
@@ -1645,6 +1645,39 @@ def test_to_with_tensor(self):
...
+        self.assertNotEqual(a.requires_grad, a.to(requires_grad=True))
+        self.assertEqual(a.requires_grad, a.to(requires_grad=False).requires_grad)
```
Comment (apaszke) [inline test/test_torch.py]: Can we make it `assertIs(a, a.to(requires_grad=False))`?

File: test/test_torch.py [hunk 3]
```diff
@@ -1645,6 +1645,39 @@ def test_to_with_tensor(self):
...
+        b = torch.tensor(5., requires_grad=True) * 5
+        self.assertRaises(RuntimeError, lambda: b.to(requires_grad=False))
+        self.assertEqual(b, b.to(requires_grad=True))
+        self.assertEqual(b.requires_grad, b.to(requires_grad=True).requires_grad)
```
Comment (apaszke) [inline test/test_torch.py]: Same here. Can this be `assertIs` instead?

File: tools/autograd/templates/python_variable_methods.cpp [hunk 4]
```diff
@@ -528,29 +529,56 @@ static PyObject * THPVariable_storage_type(PyObject* self, PyObject* arg)
...
+  } else if (requires_grad) {
+    auto res = make_variable(self_.data(), /*requires_grad=*/true);
+    res.set_version_counter(self_.version_counter());
+    return res;
+  } else {
+    if (!self_.is_leaf()) {
+      throw std::runtime_error(requires_grad_leaf_error(requires_grad));
+    }
+    self_.set_requires_grad(requires_grad);
+    return self_;
```
Comment (apaszke) [inline tools/autograd/templates/python_variable_methods.cpp]: This function is supposed to never modify the original argument, but you do it here.

File: tools/autograd/templates/python_variable_methods.cpp [hunk 5]
```diff
@@ -528,29 +529,56 @@ static PyObject * THPVariable_storage_type(PyObject* self, PyObject* arg)
...
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
-    "to(Device device, ScalarType dtype=None)",
-    "to(ScalarType dtype)",
-    "to(Tensor other)",
+    "to(*, bool requires_grad)",
+    "to(Device device, ScalarType dtype=None, *, bool requires_grad=None)",
+    "to(ScalarType dtype, *, bool requires_grad=None)",
+    "to(Tensor other)",  // lacks requires_grad because the default is ambiguous between self and other
```
Comment (apaszke) [inline tools/autograd/templates/python_variable_methods.cpp]: I can't see how not adding `requires_grad` here makes it unambiguous. The only thing you removed is the ability to specify it explicitly, but this actually never would be ambiguous anyway. I think we should just default to whatever the `requires_grad` of `self` is (which is what this implementation does I think), and use the given value otherwise. WDYT?

File: tools/autograd/templates/python_variable_methods.cpp [hunk 6]
```diff
@@ -528,29 +529,56 @@ static PyObject * THPVariable_storage_type(PyObject* self, PyObject* arg)
...
+// 3) arg is false, self requires_grad -> valid if self is a leaf.
+static Variable as_requires_grad(Tensor self, bool requires_grad) {
+  auto& self_ = static_cast<torch::autograd::Variable&>(self);
+  auto self_requires_grad = self_.requires_grad();
+  if (self_requires_grad == requires_grad) {
+    return self_;
+  } else if (requires_grad) {
+    auto res = make_variable(self_.data(), /*requires_grad=*/true);
+    res.set_version_counter(self_.version_counter());
+    return res;
```
Comment (apaszke) [inline tools/autograd/templates/python_variable_methods.cpp]: This can still have some undesirable effects if you create two independent things that `require_grad` (in-place modifications to one of them don't carry to the gradient of the other). Eh, I think we can merge this and have a big warning saying that you shouldn't do it more than once, and I'll need to think about this a bit more.

File: tools/autograd/templates/python_variable_methods.cpp (hunk 5 above)
Comment (zou3519) [reply tools/autograd/templates/python_variable_methods.cpp]: Having a `self.to(other, requires_grad=???)` makes the behavior a different from the other functions we have that have a requires_grad keyword:
- `tensor.new_zeros(...)` (returns a tensor with requires_grad=False by default) and 
- `torch.zeros_like(tensor)` (returns a tensor with requires_grad=False by default). 

Which is okay but makes me feel a uneasy

Comment (gchanan) [reply tools/autograd/templates/python_variable_methods.cpp]: Maybe the right way to think about it is x.to(*params, requires_grad=r) is equivalent to x.to(*params).to(requires_grad=r).

File: tools/autograd/templates/python_variable_methods.cpp (hunk 6 above)
Comment (gchanan) [reply tools/autograd/templates/python_variable_methods.cpp]: I'm a little dubious about adding new unsafe APIs.  We could just have a separate function that is always inplace `requires_grad_()` or `requires_grad_(requires_grad=True)`.
